File: drive/src/main/java/de/mein/drive/bash/BashToolsException.java
package de.mein.drive.bash;

import java.io.IOException;
import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;

/**
 * Created by xor on 5/24/17.
 */
public class BashToolsException extends IOException {
    private List<String> lines = new ArrayList<>();

    public BashToolsException(Iterator<String> lines) {
        readLines(lines);
    }

    public BashToolsException(String line) {
        lines.add(line);
    }

    private void readLines(Iterator<String> iterator) {
        while (iterator.hasNext())
            this.lines.add(iterator.next());
    }

    public BashToolsException(Process proc) {
        Iterator<String> lines = BashTools.inputStreamToIterator(proc.getErrorStream());
        readLines(lines);
    }

    @Override
    public void printStackTrace() {
        System.err.println(toString());
        super.printStackTrace();
    }

    @Override
    public String toString() {
        StringBuilder stringBuilder = new StringBuilder(getClass().getSimpleName() + "\n");
        for (String line : lines)
            stringBuilder.append(line).append("\n");
        return stringBuilder.toString();
    }
}

File: filesync/src/main/java/de/mel/filesync/bash/BashToolsException.kt
package de.mel.filesync.bash

import de.mel.filesync.bash.BufferedIterator.BufferedStringIterator
import java.io.IOException
import java.io.InputStreamReader
import java.util.*

/**
 * Created by xor on 5/24/17.
 */
open class BashToolsException : IOException {
    /**
     * As I learned today there is an equivalent class built into kotlin.
     */
    class NotImplemented : BashToolsException("NOT:IMPLEMENTED")

    private val lines: MutableList<String> = ArrayList()

    constructor(lines: Iterator<String>) {
        readLines(lines)
    }

    constructor(line: String) {
        lines.add(line)
    }

    private fun readLines(iterator: Iterator<String>) {
        while (iterator.hasNext()) lines.add(iterator.next())
    }

    constructor(proc: Process) {
        val lines: Iterator<String> = BufferedStringIterator(InputStreamReader(proc.errorStream))
        readLines(lines)
    }

    override fun printStackTrace() {
        System.err.println(toString())
        super.printStackTrace()
    }

    override fun toString(): String {
        val stringBuilder = StringBuilder(javaClass.simpleName + "\n")
        for (line in lines) stringBuilder.append(line).append("\n")
        return stringBuilder.toString()
    }
}